File: src/main/java/com/dxfeed/api/serializers/SeriesMapping.java
// Copyright © 2023 Devexperts LLC. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.DoubleBuffer;
import com.dxfeed.api.buffers.ByteReader;
import com.dxfeed.event.option.Series;

public class SeriesMapping {
  /**
   * typedef struct dxfg_market_event_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   * } dxfg_market_event_t;
   *
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L438
   *
   * typedef struct dxfg_series_t {
   *    dxfg_market_event_t market_event;
   *    //    dxfg_indexed_event_t indexed_event;
   *    int32_t event_flags;
   *    int64_t index;
   *    int64_t time_sequence;
   *    int32_t expiration;
   *    double volatility;
   *    double call_volume;
   *    double put_volume;
   *    double put_call_ratio;
   *    double forward_price;
   *    double dividend;
   *    double interest;
   * } dxfg_series_t;
   */

  public static void toNative(Series event, ByteBuffer pBytes, DoubleBuffer pDoubles) {
    // BYTE DATA
    pBytes.writeString(event.getEventSymbol());  // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());      // 8
    pBytes.writeInt(event.getEventFlags());      // 4
    pBytes.writeLong(event.getIndex());          // 8
    pBytes.writeLong(event.getTimeSequence());   // 8
    pBytes.writeInt(event.getExpiration());      // 8

    // DOUBLE DATA
    pDoubles.write(event.getVolatility());
    pDoubles.write(event.getCallVolume());
    pDoubles.write(event.getPutVolume());
    pDoubles.write(event.getPutCallRatio());
    pDoubles.write(event.getForwardPrice());
    pDoubles.write(event.getDividend());
    pDoubles.write(event.getInterest());
  }

  public static Series fromNative(ByteReader reader) {
    Series series = new Series();
    series.setEventSymbol(reader.readString());
    series.setEventTime(reader.readLong());
    series.setEventFlags(reader.readInt());
    series.setIndex(reader.readLong());
    series.setTimeSequence(reader.readLong());
    series.setExpiration(reader.readInt());

    series.setVolatility(reader.readDouble());
    series.setCallVolume(reader.readDouble());
    series.setPutVolume(reader.readDouble());
    series.setPutCallRatio(reader.readDouble());
    series.setForwardPrice(reader.readDouble());
    series.setDividend(reader.readDouble());
    series.setInterest(reader.readDouble());
    return series;
  }
}

File: src/main/c/jni-lib/src/events/SeriesMapping.cpp
// Copyright © 2023 Devexperts LLC. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

#include "dxfeed/events/SeriesMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"

namespace dxfeed::jni {
  dxfg_series_t* SeriesMapping::toSeries(ByteReader& reader) {
    auto* series = new dxfg_series_t();
    series->market_event.event_type.clazz = DXFG_EVENT_SERIES;
    series->market_event.event_symbol = reader.readString();
    series->market_event.event_time = reader.readLong();
    series->event_flags = reader.readInt();
    series->index = reader.readLong();
    series->time_sequence = reader.readLong();
    series->expiration = reader.readInt();

    series->volatility = reader.readDouble();
    series->call_volume = reader.readDouble();
    series->put_volume = reader.readDouble();
    series->put_call_ratio = reader.readDouble();
    series->forward_price = reader.readDouble();
    series->dividend = reader.readDouble();
    series->interest = reader.readDouble();
    return series;
  }

  void SeriesMapping::fromSeries(dxfg_series_t* series, ByteWriter& writer) {
    writer.writeString(series->market_event.event_symbol);
    writer.writeInt64_t(series->market_event.event_time);
    writer.writeInt32_t(series->event_flags);
    writer.writeInt64_t(series->index);
    writer.writeInt64_t(series->time_sequence);
    writer.writeInt32_t(series->expiration);

    writer.writeDouble(series->volatility);
    writer.writeDouble(series->call_volume);
    writer.writeDouble(series->put_volume);
    writer.writeDouble(series->put_call_ratio);
    writer.writeDouble(series->forward_price);
    writer.writeDouble(series->dividend);
    writer.writeDouble(series->interest);
  }
}